pkg/providers/amifamily/ami.go

```diff
@@ -94,69 +94,86 @@ func (p *Provider) KubeServerVersion(ctx context.Context) (string, error) {
 // If AMI overrides are specified in the AWSNodeTemplate, then only those AMIs will be chosen.
-func (p *Provider) Get(ctx context.Context, nodeTemplate *v1alpha1.AWSNodeTemplate, instanceTypes []*cloudprovider.InstanceType, amiFamily AMIFamily) (map[string][]*cloudprovider.InstanceType, error) {
-	kubernetesVersion, err := p.KubeServerVersion(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("getting kubernetes version %w", err)
-	}
+func (p *Provider) MapInstanceTypes(ctx context.Context, nodeTemplate *v1alpha1.AWSNodeTemplate, instanceTypes []*cloudprovider.InstanceType, options *Options) (map[string][]*cloudprovider.InstanceType, error) {
 	amiIDs := map[string][]*cloudprovider.InstanceType{}
-	amiRequirements, err := p.getAMIRequirements(ctx, nodeTemplate)
+	amiRequirements, err := p.Get(ctx, nodeTemplate, options)
 	if err != nil {
 		return nil, err
 	}
-	if len(amiRequirements) > 0 {
-		// Iterate through AMIs in order of creation date to use latest AMI
-		amis := sortAMIsByCreationDate(amiRequirements)
-		for _, instanceType := range instanceTypes {
-			for _, ami := range amis {
-				if err := instanceType.Requirements.Compatible(amiRequirements[ami]); err == nil {
-					amiIDs[ami.AmiID] = append(amiIDs[ami.AmiID], instanceType)
-					break
-				}
+	amis := lo.Keys(amiRequirements)
+
+	if amis[0].CreationDate != "" {
+		amis = sortAMIsByCreationDate(amis)
+	} else {
+		sort.Slice(amis, func(i, j int) bool {
+			return amis[i].Name >= amis[j].Name
+		})
+	}
+	for _, instanceType := range instanceTypes {
+		for _, ami := range amis {
+			if err := instanceType.Requirements.Compatible(amiRequirements[ami]); err == nil {
+				amiIDs[ami.AmiID] = append(amiIDs[ami.AmiID], instanceType)
+				break
 			}
 		}
-		if len(amiIDs) == 0 {
-			return nil, fmt.Errorf("no instance types satisfy requirements of amis %v,", lo.Keys(amiRequirements))
+	}
+	if len(amiIDs) == 0 {
+		return nil, fmt.Errorf("no instance types satisfy requirements of amis %v,", lo.Keys(amiRequirements))
+	}
+
+	return amiIDs, nil
+}
+
+func (p *Provider) Get(ctx context.Context, nodeTemplate *v1alpha1.AWSNodeTemplate, options *Options) (map[AMI]scheduling.Requirements, error) {
+	var err error
+	amiRequirements := map[AMI]scheduling.Requirements{}
+	if len(nodeTemplate.Spec.AMISelector) == 0 {
+		amiRequirements, err = p.getDefaultAMIFromSSM(ctx, nodeTemplate, options)
+		if err != nil {
+			return nil, err
 		}
 	} else {
-		for _, instanceType := range instanceTypes {
-			amiID, err := p.getDefaultAMIFromSSM(ctx, amiFamily.SSMAlias(kubernetesVersion, instanceType))
+		allAmis, err := p.getAMIsFromSelector(ctx, nodeTemplate)
+		if err != nil {
+			return nil, err
+		}
+
+		requirementsHash := map[uint64]bool{}
+		amis := sortAMIsByCreationDate(lo.Keys(allAmis))
+		for _, ami := range amis {
+			hash, err := hashstructure.Hash(allAmis[ami].NodeSelectorRequirements(), hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
```

Comment:
Can we guarantee that this won't error? I think so and if so, we can use a `lo.Must()` here